p2p/src/types.rs

```diff
@@ -0,0 +1,473 @@
+//! Shared address types for p2p networking.
+
+use bytes::{Buf, BufMut};
+use commonware_codec::{EncodeSize, Error as CodecError, FixedSize, Read, ReadExt, Write};
+use commonware_runtime::{Error as RuntimeError, Resolver};
+use commonware_utils::{Hostname, IpAddrExt};
+use std::net::{IpAddr, SocketAddr};
+
+const INGRESS_SOCKET_PREFIX: u8 = 0;
+const INGRESS_DNS_PREFIX: u8 = 1;
+
+const ADDRESS_SYMMETRIC_PREFIX: u8 = 0;
+const ADDRESS_ASYMMETRIC_PREFIX: u8 = 1;
+
+/// What we dial to connect to a peer.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Ingress {
+    /// IP-based ingress address.
+    Socket(SocketAddr),
+    /// DNS-based ingress address.
+    Dns {
+        /// Hostname to resolve.
+        host: Hostname,
+        /// Port to connect to.
+        port: u16,
+    },
+}
+
+impl Ingress {
+    /// Returns the port number for this ingress address.
+    pub const fn port(&self) -> u16 {
+        match self {
+            Self::Socket(addr) => addr.port(),
+            Self::Dns { port, .. } => *port,
+        }
+    }
+
+    /// Returns the IP address if this is a Socket variant.
+    pub const fn ip(&self) -> Option<IpAddr> {
+        match self {
+            Self::Socket(addr) => Some(addr.ip()),
+            Self::Dns { .. } => None,
+        }
+    }
+
+    /// Returns whether this ingress address is allowed given the configuration.
+    ///
+    /// - `Socket` addresses must have a global IP (or `allow_private_ips` must be true).
+    /// - `Dns` addresses are allowed only if `allow_dns` is `true`.
+    ///
+    /// Note: For `Dns` addresses, private IP checks are performed after resolution in
+    /// [`resolve_filtered`](Self::resolve_filtered).
+    pub fn is_valid(&self, allow_private_ips: bool, allow_dns: bool) -> bool {
+        match self {
+            Self::Socket(addr) => allow_private_ips || IpAddrExt::is_global(&addr.ip()),
+            Self::Dns { .. } => allow_dns,
+        }
+    }
+
+    /// Resolve this ingress address to a socket address.
+    ///
+    /// For `Socket` variants, returns the address directly.
+    /// For `Dns` variants, performs DNS resolution and returns the first resolved IP.
```

Comment:
Should we return an iterator over the ips here? Folks may want to try each until one connects. Some of the more complicate infra setups of our customers might have round robin DNS setups.